source/common/network/tcp_keepalive_option_impl.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include <netinet/in.h>
+#include <netinet/ip.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+
+#include "envoy/network/listen_socket.h"
+
+#include "common/common/logger.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Network {
+
+// Optional variant of setsockopt(2) optname. The idea here is that if the option is not supported
+// on a platform, we can make this the empty value. This allows us to avoid proliferation of #ifdef.
+typedef absl::optional<int> SocketOptionName;
+
+#ifdef SO_KEEPALIVE
+#define ENVOY_SOCKET_SO_KEEPALIVE Network::SocketOptionName(SO_KEEPALIVE)
+#else
+#define ENVOY_SOCKET_SO_KEEPALIVE Network::SocketOptionName()
+#endif
+
+#ifdef TCP_KEEPCNT
+#define ENVOY_SOCKET_TCP_KEEPCNT Network::SocketOptionName(TCP_KEEPCNT)
+#else
+#define ENVOY_SOCKET_TCP_KEEPCNT Network::SocketOptionName()
+#endif
+
+#ifdef TCP_KEEPIDLE
+#define ENVOY_SOCKET_TCP_KEEPIDLE Network::SocketOptionName(TCP_KEEPIDLE)
+#elif TCP_KEEPALIVE // OS X I think
```

Comment:
OK. Do i need to support OSes other than OSX and Linux? I think there's at least one where the options are in ms rather than s, and Windows seems to use a single option with a struct, which would make things a little messier.

Reply:
Linux is required. Macos is optional. I don't think envoy builds on any other platforms.